src/samplers/mh.jl

```diff
@@ -0,0 +1,249 @@
+      push!(spl.info[:accept_his], true)
+    else                      # rejected
+      push!(spl.info[:accept_his], false)
+      vi[spl] = old_θ         # reset Θ
+      setlogp!(vi, old_logp)  # reset logp
+    end
+
+    vi
+  end
+end
+
+function sample(model::Function, alg::MH;
+                chunk_size=CHUNKSIZE,     # set temporary chunk size
+                save_state=false,         # flag for state saving
+                resume_from=nothing,      # chain to continue
+                reuse_spl_n=0,            # flag for spl re-using
+                )
+
+  default_chunk_size = CHUNKSIZE  # record global chunk size
+  setchunksize(chunk_size)        # set temp chunk size
+
+  spl = reuse_spl_n > 0 ?
+        resume_from.info[:spl] :
+        Sampler(alg)
+  alg_str = "MH"
+
+  # Initialization
+  time_total = zero(Float64)
+  n = reuse_spl_n > 0 ?
+      reuse_spl_n :
+      alg.n_iters
+  samples = Array{Sample}(n)
+  weight = 1 / n
+  for i = 1:n
+    samples[i] = Sample(weight, Dict{Symbol, Any}())
+  end
+
+  vi = resume_from == nothing ?
+       model() :
+       deepcopy(resume_from.info[:vi])
+
+  if spl.alg.gid == 0
+    runmodel(model, vi, spl)
+  end
+
+  # MH steps
+  if PROGRESS spl.info[:progress] = ProgressMeter.Progress(n, 1, "[$alg_str] Sampling...", 0) end
+  for i = 1:n
+    dprintln(2, "$alg_str stepping...")
+
+    time_elapsed = @elapsed vi = step(model, spl, vi, i == 1)
+    time_total += time_elapsed
+
+    if spl.info[:accept_his][end]     # accepted => store the new predcits
+      samples[i].value = Sample(vi, spl).value
+    else                              # rejected => store the previous predcits
+      samples[i] = samples[i - 1]
+    end
+    samples[i].value[:elapsed] = time_elapsed
+
+    if PROGRESS ProgressMeter.next!(spl.info[:progress]) end
+  end
+
+  println("[$alg_str] Finished with")
+  println("  Running time        = $time_total;")
+  accept_rate = sum(spl.info[:accept_his]) / n  # calculate the accept rate
+  println("  Accept rate         = $accept_rate;")
+
+  setchunksize(default_chunk_size)      # revert global chunk size
+
+  if resume_from != nothing   # concat samples
+    unshift!(samples, resume_from.value2...)
+  end
+  c = Chain(0, samples)       # wrap the result by Chain
+  if save_state               # save state
+    save!(c, spl, model, vi)
+  end
+
+  c
+end
+
+function rand_truncated(dist, lowerbound, upperbound)
+    notvalid = true
+    x = 0.0
+    while (notvalid)
+        x = rand(dist)
+        notvalid = ((x < lowerbound) | (x > upperbound))
+    end
+    return x
+end
+
+assume(spl::Sampler{MH}, dist::Distribution, vn::VarName, vi::VarInfo) = begin
+    if isempty(spl.alg.space) || vn.sym in spl.alg.space
+      vi.index += 1
+      old_val = getval(vi, vn)[1]
+
+      if ~haskey(vi, vn)
+        r = rand(dist)
+        push!(vi, vn, r, dist, spl.alg.gid)
+        spl.info[:proposal_ratio] += (logpdf(dist, old_val) - logpdf(dist, r))
```

Comment:
Might be a silly question: if `haskey(vi,vn)` equails to `false`, wouldn't `logpdf(dist, old_val)` throw an error since `old_val` does not yet exist?

Reply:
I think you are right ! Not sure if we can handle this case. `MH` may be restricted to a deterministic number of random variables (similar to `HMC`).
@twgr would know whether Metropolis-Hasting can handle stochastic existence of random variables ? i.e. should the set of random variables that is updated be fixed (as for `HMC` ?).

Reply:
MH can handle stochastic existence of variables if you're careful - it is more difficult than usually to ensure your proposal leads to an ergodic Markov chain though as you need to make sure you can get to everywhere from everywhere.  The easiest way to do this is to have your proposal be a mixture of sampling from the prior and something else. There is also all the LMH stuff that is a particular case of MH with random existence.  As an example of the stuff that can go wrong here, there were a lot of things wrong with the original Wingate paper, for which I give an explanation in my thesis.

Reply:
I would vote for the option of using a mixture of prior and Gaussian (or other proposals) since this would only require a small modification here. For example, we could sample `old_val` from the prior if it does not exist yet; then for following steps, we could flip a coin to decide to propose from the prior or user-specified proposals.

Reply:
therefore something like `old_val = haskey(vi, vn) ? vi[vn] : rand(dist)`

Reply:
that looks good to me.

Reply:
You might want to include some part of the proposal that allows single elements to be added or dropped -- though it provides ergodicity, relying on a prior component in the proposal to change the dimension can lead to very slow mixing for some models.